### PR TITLE
New lint: suggest `ptr::read` instead of `mem::replace(..., uninitialized())`

### PR DESCRIPTION
**Pull request by esamudera**
resolves: #5575 

changelog: improvements to `MEM_REPLACE_WITH_UNINIT`:
- add a new test case in `tests/ui/repl_uninit.rs` to cover the case of replacing with `mem::MaybeUninit::uninit().assume_init()`.
- modify the existing `MEM_REPLACE_WITH_UNINIT` when replacing with `mem::uninitialized` to suggest using `ptr::read` instead.
- lint with `MEM_REPLACE_WITH_UNINIT` when replacing with `mem::MaybeUninit::uninit().assume_init()`

### CODE REVIEW

File: clippy_lints/src/mem_replace.rs [hunk 1]
```diff
@@ -135,33 +135,66 @@ fn check_replace_option_with_none(cx: &LateContext<'_, '_>, src: &Expr<'_>, dest
     }
 }
 
-fn check_replace_with_uninit(cx: &LateContext<'_, '_>, src: &Expr<'_>, expr_span: Span) {
-    if let ExprKind::Call(ref repl_func, ref repl_args) = src.kind {
+fn check_replace_with_uninit(cx: &LateContext<'_, '_>, src: &Expr<'_>, dest: &Expr<'_>, expr_span: Span) {
+    if let ExprKind::MethodCall(method_path, _, method_args) = src.kind {
```
Comment (phansch) [inline clippy_lints/src/mem_replace.rs:139]: This `if let` can be moved down into the `if_chain` to remove one level of indentation

File: clippy_lints/src/mem_replace.rs [hunk 2]
```diff
@@ -135,33 +135,66 @@ fn check_replace_option_with_none(cx: &LateContext<'_, '_>, src: &Expr<'_>, dest
...
         if_chain! {
+            // check if replacement is `maybe_uninit::uninit().assume_init()`:
+            if method_path.ident.name == sym!(assume_init);
+            if ! method_args.is_empty();
```
Comment (phansch) [inline clippy_lints/src/mem_replace.rs:143]:
```suggestion
            if !method_args.is_empty();
```

File: clippy_lints/src/mem_replace.rs [hunk 3]
```diff
@@ -135,33 +135,66 @@ fn check_replace_option_with_none(cx: &LateContext<'_, '_>, src: &Expr<'_>, dest
     }
 }
 
-fn check_replace_with_uninit(cx: &LateContext<'_, '_>, src: &Expr<'_>, expr_span: Span) {
-    if let ExprKind::Call(ref repl_func, ref repl_args) = src.kind {
+fn check_replace_with_uninit(cx: &LateContext<'_, '_>, src: &Expr<'_>, dest: &Expr<'_>, expr_span: Span) {
+    if let ExprKind::MethodCall(method_path, _, method_args) = src.kind {
         if_chain! {
+            // check if replacement is `maybe_uninit::uninit().assume_init()`:
+            if method_path.ident.name == sym!(assume_init);
```
Comment (oli-obk) [inline clippy_lints/src/mem_replace.rs:142]: If you want you could also make this function a diagnostic item in rustc and then reference it directly this way (instead of checking the name and doing a diagnostic item check for the `MaybeUninit` struct.

File: clippy_lints/src/mem_replace.rs (hunk 1 above)
Comment (esamudera) [reply clippy_lints/src/mem_replace.rs]: Okay, I have moved the `MethodCall` check inside the `if_chain`

File: clippy_lints/src/mem_replace.rs (hunk 2 above)
Comment (esamudera) [reply clippy_lints/src/mem_replace.rs]: This formatting typo has been fixed

File: clippy_lints/src/mem_replace.rs (hunk 3 above)
Comment (esamudera) [reply clippy_lints/src/mem_replace.rs]: Okay, I'm not very familiar with diagnostic items, but reading from the source code, it should be the API described in this issue right? https://github.com/rust-lang/rust-clippy/issues/5393

I will read more on this, are there any other references I should also read on diagnostic item API? Thanks :)

Comment (oli-obk) [reply clippy_lints/src/mem_replace.rs]: I was sure I wrote docs, but I can't find anything. I apologize. I'll add some docs to the rustc guide.

TLDR: slap `#[rustc_diagnostic_item = "foobar")]` on a function or type declaration in rustc in order for `cx.tcx.is_diagnostic_item(sym::foobar, some_def_id)` to be true  if `some_def_id` is the `DefId` of said item. You also need to add `foobar` to the list in https://github.com/rust-lang/rust/blob/118b50524b79e565f017e08bce9b90a16c63634f/src/librustc_span/symbol.rs#L111

Comment (esamudera) [reply clippy_lints/src/mem_replace.rs]: So I checked the Rust source code, and found that `#[rustc_diagnostic_item = "assume_init"]` is already implemented: https://github.com/rust-lang/rust/blob/master/src/libcore/mem/maybe_uninit.rs#L495

I then implemented the diagnostic_item check and wow you're right, the code is greatly simplified and much more readable now. 

With this, I guess this implementation is pretty much done right? If yes, I can rebase everything into one and remove the draft status from this PR :+1:

**Comment by esamudera** [comment]
Hi @oli-obk, just want to follow up on this PR in case my previous comment got buried. I will quote my previous comment:

> So I checked the Rust source code, and found that `#[rustc_diagnostic_item = "assume_init"]` is already implemented: https://github.com/rust-lang/rust/blob/master/src/libcore/mem/maybe_uninit.rs#L495
> 
> I then implemented the diagnostic_item check and wow you're right, the code is greatly simplified and much more readable now.
> 
> With this, I guess this implementation is pretty much done right? If yes, I can rebase everything into one and remove the draft status from this PR 👍

And this is the commit: https://github.com/rust-lang/rust-clippy/pull/5695/commits/a62ae7ca9dc9f8a37727da7a1714f2c9c1cb8837

Thanks!

File: clippy_lints/src/mem_replace.rs (hunk 3 above)
Comment (oli-obk) [reply clippy_lints/src/mem_replace.rs]: yes, this PR lgtm now! Thanks
